### ISSUE TITLE
Android build doesn't use proper paths

### ISSUE THREAD

**Issue by ShadowNinja** [minetest/minetest#2097]
Android provides some folders to apps for them to store there data -- which Minetest ignores, using `/sdcard/Minetest/` instead.  This means that when uninstalling Minetest the folder is left behind, wasting space.  It also means that Android's "Clear data" and "Clear cache" functions don't work.  Aditionally, Android 4.4+ doesn't require the `READ_EXTERNAL_STORAGE` and `WRITE_EXTERNAL_STORAGE` permissions to write to these locations, but does for access to the sdcard root.

  * Data files should be stored in the files path ([getDir()](http://developer.android.com/reference/android/content/Context.html#getDir(java.lang.String, int)) / [getFilesDir()](http://developer.android.com/reference/android/content/Context.html#getFilesDir()) / [getExternalFilesDir()](http://developer.android.com/reference/android/content/Context.html#getExternalFilesDir(java.lang.String))).
  * Cache files should be stored in the cache path ([getCacheDir()](http://developer.android.com/reference/android/content/Context.html#getCacheDir()) / [getExternalCacheDir()](http://developer.android.com/reference/android/content/Context.html#getExternalCacheDir())).

**Comment by MoNTE48** [comment]
Good idea!
Or use /Android/obb/package_name (/Android/data/package_name)

**Comment by ShadowNinja** [comment]
@MoNTE48 The OBB folder is only for storing OBBs.  A subdirectory in the data path is given by the functions I mentioned.  Also, you shouldn't use static paths, they change often.  Call the relevant function to get the path.

**Comment by MoNTE48** [comment]
OK.  Your  ideas good.
wait realised.

**Comment by sapier** [comment]
this is not a bug but a missing feature.
BTW it's irrelevant if android 4.x doesn't require something any more because we support from 2.3 so if something is required there we still need it.

**Comment by MoNTE48** [comment]
Hmm... Real Racing game download cache files (+3000) on Android/data/package_name.

**Comment by sapier** [comment]
Is that game open source so we can have a look how they do it?
btw recent irrlicht does have some asset support maybe we can skip copying textures. mods builtin to sdcard.

**Comment by MoNTE48** [comment]
No...
More game copy game data on
/sdcard/game(s)/folder.

**Comment by ShadowNinja** [comment]
@sapier I consider this a bug because it's like a storage memory leak -- even after uninstalling you have wasted memory.

You wouldn't pass file descriptors to the engine, just the paths.

**Comment by MoNTE48** [comment]
I can write that after you uninstall the game folder minetest just deleted automatically.

**Comment by sapier** [comment]
ShadowNinja are you sure it'd be a good idea to assume the "path" provided by that api is a actual filesystem path. The api doesn't actually provide a "get path" function and implementation details may change. I admitt it might work right now but It'd not be first time android changed paths and just provides the correct ones per api. One example for "virtual" paths are assets for example.
A possible situation why direct access to those paths might be not possible any longer would be encrypting those folders. Java would handle it internally but native code couldn't.

I still consider this to be a nice to have feature so if someone implements it and it actually works from 2.3 to current android version I'd not wait but merge it ;-) Yet I believe there's quite a lot of trouble we can cause by it, this isn't a trivial thing to do.

**Comment by ShadowNinja** [comment]
@sapier They return File objects.  File objects aren't actual file descriptors, they're just a wrapper around a path string with some helper functions.  See the documentation here: http://developer.android.com/reference/java/io/File.html.  You can pass the return value of `getAbsolutePath()` to Minetest for `path_user` / `path_share` and a new `path_cache`.

I don't know how android handles encryption, but if you can access the `Minetest/` folder you should be able to access the path returned by `getExternalFilesDir()`.  It also seems really odd for native applications to have to manually decrypt files, especially since there are so many native apps (just about any 3D game has to be native for speed).

**Comment by est31** [comment]
Here is a [forum post](https://forum.minetest.net/viewtopic.php?f=6&t=12374) describing an user who wants to install minetest over a broken fork, but fails because of this issue.

**Comment by est31** [comment]
Everything I've seen about android encryption was dm-crypt based, using the mount system, thus transparent to apps.

**Comment by stujones11** [comment]
I am not sure what is controversial here, I think that everyone agrees that we should be using the proper application data path like pretty much every other app does. Since we no longer support android 2.3 sapier's comments are no longer really valid. I have had this working before on another test build.

Doing this may also solve the problems we will face when users eventually upgrade from 0.4.x to 5.0.0 as this currently requires removal or renaming of the `Minetest` directory, otherwise the game and/or builtin scripts do not get updated properly, I forget the exact details.

The only downside is that users would then need manually copy existing singleplayer worlds to the new install location. We could perhaps include a simple migration tool, however, that would need to be done in java so not something I'd be able to help with.

**Comment by paramat** [comment]
Yes that label wasn't very suitable as the focus here is the bug not the solution.

**Comment by MoNTE48** [comment]
I raise this question and give a brief explanation. My previous fix and this PR [https://github.com/minetest/minetest/pull/9900] give the last possible fix for Android 10.

Starting with Android 11 (probably in late summer or fall), none of this will work anymore! Google will completely ban apps from writing to /sdcard. Do not give your solutions, they will work only for certain types of files (photos, videos and some types of files). Native code either will not be able to work correctly, or will require serious hacks, followed by a significant drop in I/O performance!
Solution 1) /sdcard/Android/data/app_name/files
Solution 2) /data/data/app_name/files
Solution 3) Indicate Minetest not supporting Android 11 and higher.

**Comment by MoNTE48** [comment]
@sfan5, well, and yes and no. There are 2 scenarios.
1) We can use API 29 until Google blocks sending new APKs with this target API.
2) Android 11 can force a block on direct access to /sdcard. Then Minetest will simply break for those who update Android. We need to wait for the release of the first beta version.

**Comment by rubenwardy** [comment]
This will need to be fixed by November 2021 at the absolute latest, otherwise data will be lost by users

**Comment by weather-tracker** [comment]
Looks like you can still read write to external storage but not spacific dir

Starting in Android 11, apps cannot create their own app-specific directory on external storage. To access the directory that the system provides for your app, call getExternalFilesDirs().

https://developer.android.com/about/versions/11/privacy/storage
